### ISSUE TITLE
[TRACKING] MyOwnFreeHost Integration

### ISSUE THREAD

**Issue by evrifaessa** [boxbilling/boxbilling#766]
So, I'll be working on a possible MyOwnFreeHost integration. Wanted to create an issue just to track the process, I'll take notes here, and inform those who are interested also from here.

I'll be using @HansAdema's [InfinityFreeHosting/mofh-client](https://github.com/InfinityFreeHosting/mofh-client).

From what I expect, these functions will be available:
- getLoginUrl()
- getResellerLoginUrl()
- testConnection()
- createAccount()
- suspendAccount()
- unsuspendAccount()
- changeAccountPassword()

These **won't** be available, and won't function when being called:
- cancelAccount()
- changeAccountPackage()
- changeAccountUsername()
- changeAccountDomain()
- changeAccountIp()
- _getResourcePlan()
- _getDnsTemplate()

**Comment by HansAdema** [comment]
Glad to see my work is being turned into something that's more accessible to people! And glad to see that Boxbilling is being developed at all once again.

Just adding a few thoughts here regarding the overall idea:

- One of the annoying things about the MOFH API is that you need to generate a username yourself within very tight constraints, and get another username back from MOFH when creating the account. You'll need to store both. The username generated by you is used for most API calls, the username generated by MOFH is most relevant for the user. This is a bit different from the classic cPanel style of credential management, which tends to be problematic with least-common-denominator generic hosting platform adapters.
- Listening to the webhooks is extremely valuable to keep the local database in sync with the MOFH platform. Since MOFH quite frequently suspends accounts on their own (and sometimes reactivates them too), these notifications are very useful.

As for your specific functions:

> - getLoginUrl()
> - getResellerLoginUrl()

The API (and consequently the API client) doesn't offer this. For InfinityFree, I created a self-submitting form using the example from the Remote Client Login form from the reseller panel.

> - testConnection()

The API doesn't have a ping function, but using the `getdomainuser` function with a dummy domain should allow you to catch any authentication errors.

> - cancelAccount()

You could implement this by suspending the account with the reason `REQUESTED`. It's not entirely the same as terminating the account, but it may be preferable to having a dangling function.

**Comment by evrifaessa** [comment]
@HansAdema, thanks for the reply :)

> The API (and consequently the API client) doesn't offer this. For InfinityFree, I created a self-submitting form using the example from the Remote Client Login form from the reseller panel.

For now, it only returns 'https://panel.myownfreehost.net/' as the reseller panel URL, and the typical vPanel URL for the login link. Sadly, no auto-login stuff like cPanel or Plesk at all.

> The API doesn't have a ping function, but using the getdomainuser function with a dummy domain should allow you to catch any authentication errors.
That's what I exactly did at the beginning. We're using the same method for the Kloxo module already.

I've applied for a MOFH account, but still waiting for iFastNet to activate it. Until then, I'm just coding the module with no proper tests at all. When I get the key pairs for the API, I'll start testing what I did.
